service/controller/resource/terminateunhealthynode/create.go

```diff
@@ -0,0 +1,177 @@
+		}
+
+		detectorService, err = detector.NewDetector(detectorConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	nodesToTerminate, err := detectorService.DetectBadNodes(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if len(nodesToTerminate) > 0 {
+		for _, n := range nodesToTerminate {
+			err := r.terminateNode(ctx, n, cr)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+
+		// reset tick counters on all nodes in cluster to have a graceful period after terminating nodes
+		err := detectorService.ResetTickCounters(ctx)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", "resetting tick node counters on all nodes in tenant cluster")
+	}
+
+	return nil
+}
+
+func (r *Resource) getTenantClusterClient(ctx context.Context, cluster *capiv1alpha3.Cluster) (client.Client, error) {
+	var k8sClient k8sclient.Interface
+	{
+		restConfig, err := r.tenantRestConfigProvider.NewRestConfig(ctx, key.ClusterID(cluster), cluster.Spec.ControlPlaneEndpoint.Host)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		k8sClient, err = k8sclient.NewClients(k8sclient.ClientsConfig{
+			Logger:     r.logger,
+			RestConfig: rest.CopyConfig(restConfig),
+		})
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	return k8sClient.CtrlClient(), nil
+}
+
+func (r *Resource) terminateNode(ctx context.Context, node corev1.Node, cluster capiv1alpha3.Cluster) error {
+	if node.Labels["role"] != "worker" {
+		// We can only terminate workers in azure.
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Termination of master nodes is not supported on Azure")
+		return nil
+	}
+
+	vmssClient, err := r.azureClientsFactory.GetVirtualMachineScaleSetsClient(ctx, cluster.ObjectMeta)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Scale VMSS up by one.
+	var vmssName string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Retrieving AzureMachinePool CR")
+		amp := capzexpv1alpha3.AzureMachinePool{}
+		err := r.ctrlClient.Get(ctx, client.ObjectKey{Namespace: cluster.Namespace, Name: node.Labels[label.MachinePool]}, &amp)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Retrieved AzureMachinePool CR")
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Retrieving VMSS")
+		vmssName = key.NodePoolVMSSName(&amp)
+		vmss, err := vmssClient.Get(ctx, cluster.Name, vmssName)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "level", "debug", "message", "Retrieved VMSS")
+
+		newCapacity := *vmss.Sku.Capacity + 1
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("Scaling up VMSS %q to %d replicas", vmssName, newCapacity))
+
+		update := compute.VirtualMachineScaleSetUpdate{
+			Sku: &compute.Sku{
+				Capacity: &newCapacity,
+			},
+		}
+
+		_, err = vmssClient.Update(ctx, cluster.Name, vmssName, update)
```

Comment:
Does termination below decrease the VMSS capacity back to desired?

Reply:
deleting one node automatically decreses the size of VMSS, hence the scale up above